### ISSUE TITLE
Processing xml:ids for stage directions in modern texts

### ISSUE THREAD

**Issue by mahaylagalliford** [projectLEMDO/lemdoIssues#136]
We want to add to the processing for the ASSP system to include stage directions: ASSD. Act and scene will be processed as usual, while the SD pulls from the xml:id on stage elements. We do not want speech included in canonical references for stage directions (see rationale). For example, xml:ids see emd1HW_M (act 1, scene 1 and act 3, scene 1). We have a `<ptr>` from emd1HW_M_annotation  linked to SD with xml:id="emd1HW_M_a3_s1_sd21". For this example, we want its render as: A3 Sc1 SD21

Rationale: From my JCURA research I learned that stage directions are paratextual material and that we need to be able to link to them in modern texts. LEMDO has not yet resolved the question of how to make stage directions citable and linkable in the modern texts. I recommend a citation system similar to our ASSP system, however, since stage directions are paratextual, they are contained to the scene and are independent from the surrounding text. Therefore, I suggest that, within each scene, we give every stage direction an xml:id numbered sequentially (sd1) and that we start the numbering process over again with each new scene.

**Comment by martindholmes** [comment]
What if you have a pointers which spans from a stage-direction to a subsequent speech?

**Comment by martindholmes** [comment]
Expanding on this a bit, I see two problems, or at least issues that we'll need to clarify before we can proceed:

1. Unless stage direction numbers/ids appear in the margin, the reader won't be able to navigate with them, in the PDF and in the online view. But because they can co-occur with the beginnings of speeches, it won't always be possible to put their identifiers in the margins in the PDF. In the H5 pdf, p.20 (A2 Sc1 Sp20) you'll see one of many examples of this. In the online version, because we always have a linebreak following the speaker name, this is not so much of a problem, but it's a killer for the PDF. We'll have the same problem for any stage direction that co-occurs with a line number. I can't see a solution to this, other than to prioritize one and discard the other, which will be mighty confusing for readers.
2. It's not clear what should happen if an editor annotates a span which begins or ends inside a stage direction but spans outside it. Should we just assume that because non-SD content is involved, we should ignore the SD component in the reference?

**Comment by LEMDO-PM** [comment]
1. I think that we discussed not adding SD numbers in the PDF for that reason. I'll ping @JanelleJenstad to confirm that. Could we leave the processing as-is for the PDF and add processing for `@xml:id` on `<stage>` in our digital output?
2. Yes, I think that that's fair. If we didn't add SD numbers in PDFs, does this question still stand for annotations? For parenthetical citations, could people simply add two `<ptr>` elements with an en-dash between so that it would render as, for example, A3 Sc1 SD21-A3 Sc1 Sp30?

**Comment by martindholmes** [comment]
@LEMDO-PM 

Re 1, yes, we can ignore these things in the PDF, but that will mean that editors will have to avoid using them entirely in any text which may form part of a PDF at some stage. That doesn't seem easily enforceable. 

Re 2, I agree that two ptrs is the way to do this sort of thing, but editors are not necessarily going to know or notice when they cross these boundaries. They're used to using a single pointer to two anchors to annotate a span (there are currently 2273 localCit pointers which point to a pair of anchors). We could have a Schematron rule which says that you can't have a single anchor inside a stage-direction -- in other words, you either point directly to the stage direction's id or you point to two anchors inside it. But there are already 62 cases of a single anchor inside a stage direction, so there would be significant remediation to be done.

**Comment by martindholmes** [comment]
@LEMDO-PM I see you removed the blueSky label, but we still don't have any practical plan for handling this. The two major problems remain:

1. Stage directions coincide with every other ASSP marker type, which means there will be collisions all over the place and no practical way to show their ids/anchors in the HTML.
2. For that reason we can't display them at all in the PDF, where page width is already a problem, and therefore any link pointing to a stage direction will be confusing for readers in the PDF. 

I think this is entirely impractical, UNLESS we force all stage directions to be block elements starting on a new line. If we can do that, then this is feasible. How do we feel about that?

**Comment by JanelleJenstad** [comment]
I have a different suggestion:

- DON'T display SD numbers in the HTML or PDF at all. It's too messy. However, rendering problems should not stop us from doing something valuable in the XML. So we should:

- GIVE <stage> elements an xml:id. They can be consecutive in the semi-dip texts from _sd1 to _sd??. In modern texts, they can be constructed as I've suggested before: emdABBR_M_a1_s1_sd1 (with no reference to speeches).

Giving them xml:ids allows us to do the following:

- Link to stage directions
- Display stage directions in a separate file (as the Oxford, New Oxford, and Oxford Middleton do), where the numbers can be displayed.
- Make it possible for other interfaces to be built in the future (for example, an interface in which all stage directions are put in an adjacent column, as some print editions have done recently)

Bottom line: The difficulty of processing into a clickable number should not stop us from assigning xml:ids.

**Comment by martindholmes** [comment]
@janelleJenstad There are already 266 stage directions with `@xml:id`, and these turn into `@id`s on the HTML output element, so they can be linked to. There are two actual pointers linking to these, one in emd1HW_M_annotation.xml and one in emdJC_TextIntro.xml. The links actually work in the output, but in the first case the link text suggests there is an error, and in the second, only the word "link" appears.

So the next question is what link text should show up for these types of links. The suggestion above for expanding the id into act, scene, stage-direction works for me; although that doesn't represent a caption or label that appears in the text, it would at least make some sense to the reader. What do you think?

**Comment by martindholmes** [comment]
Pinging @JanelleJenstad and @LEMDO-PM on this -- should we just abandon this and make people use a ref element with text to form a link when they want to point to a stage direction? It seems more straightforward, given that it's not going to happen all that often.

**Comment by LEMDO-PM** [comment]
I'll leave it to @JanelleJenstad what we want to appear as the text, but I don't want to abandon this ticket. I think that the suggestion to expand ASSp numbers for the pointers to SDs makes sense.

**Comment by JanelleJenstad** [comment]
"So the next question is what link text should show up for these types of links." 
We want A1 S1 SD1

So emd1HW_M_a1_s2_sd3 would render as A1 S2 SD3

**Comment by martindholmes** [comment]
@JanelleJenstad @LEMDO-PM So my next question is this:

Given that most stage-directions will never be linked to at all, it seems mad to add all these xml:ids to no purpose; the ASSP ids all have a purpose (identifying anchorable locations which are visible and usable in the output), but since these won't be showing at all, we presumably don't want to expand our texts by adding them all over the place just in case. So that means when an id is needed, it has to be addable in an ad-hoc basis, but since the id itself is to be used to calculate the text which is shown in the expanded link, it needs to be correct. As with ASSP ids, an editing action can render the numeric components incorrect, so they're a bit fragile. Do we impose on editors to calculate the id that a SD should have, or do we try to provide some mechanism which will calculate it for them? (I have some Schematron code that does something like that already in the Monument project.)

I'd like to make sure we get this completely right before we act on it, because the ASSP-handling code is quite complicated and this will add more complication.

**Comment by LEMDO-PM** [comment]
I think that in file where the editor wants to link to stage directions, they should number all stage directions. I have found that when editors link to one SD, they're likely to link to others, so they should put the xml:ids in place that they may end up relying on later. Editors that don't want to link to stage directions don't need to give any of them xml:ids.

I think the simplest thing to do would be to write an XSLT to number SDs based on their parent `<div>`'s xml:id and then also a Schematron rule that they must be in sequential order in the file (same as the rule for `<sp>` elements). Then editors have the attributes in place to link to what they need; there's a way to easily, quickly, and accurately number stage directions; and editors will get an error if the numbers are incorrect.

**Comment by JanelleJenstad** [comment]
We need to split this ticket into (1) what we do in the digital output, and (2) what we do in the PDF/print output. We also need to split it into (A) short-term actions (to facilitate DRE release) and (B) longer-term actions (to facilitate printing of the PDF/print output). 

Let's agree that we do NOT have to implement the PDF/print solution in the short-term. 

Short-term solution [digital only]:

- We will add xml:ids to stage directions. @LEMDO-PM has an XSLT to do this. The xml:ids will inherit the values of the parent div and number consecutively (with no reference to speech numbers).
- When editors wish to talk about a stage direction (in an annotation or a critical paratext), we will make a pointer to the xml:id of the stage direction.
- We will make a link that allows readers to jump to the stage direction.
- We will NOT display the stage direction number in the modern or semi-dip texts.
- For the text of the link, we will use this convention: A1 Sc1 SD1

Long-term solution [print solution + an enhancement of the digital solution that will already have been implemented by then]

- Digital display of modern and semi-dip texts will have a radio button that allows users to toggle between speech numbers (ASSP view in modern; SP view in semi-dip) and stage directions (ASSD view in modern; SD view in semi-dip). 
- Digital edition will have an HTML page listing all the stage directions harvested from the semi-dip text, with their SD numbers listed AND a link back to the semi-dip text.
- Digital edition will have another HTML page listing all the stage directions harvested from the modern text, with their xml:ids and ASSD numbers listed AND a link back to the modern text.
- Print edition renders ASSD numbers where pointers appear in the XML. 

- Print edition has two possible solutions to the problem of not having room to display ASSD numbers in the margins:
-- We include an Appendix with all stage directions with their ASSD numbers.
-- We include in the series introduction a note directing readers to the digital text to find stage directions by their ASSD numbers (not ideal but functional).

**Comment by JanelleJenstad** [comment]
@martindholmes Just pinging you on this ticket. @LEMDO-PM can talk it through with you. It's release-blocking for DRE because DRE editions point to stage directions from various places.

**Comment by martindholmes** [comment]
I don't think there's anything to do, is there?

**Comment by martindholmes** [comment]
I thought we already decided that sds can't be part of ASSP, any as far as I know, any link made with a ref pointing to a sd will just work, no?

**Comment by LEMDO-PM** [comment]
The issue is that people can't consistently put text in a ref pointing to stage directions if there is not something consistent to point to, and we would like people's references to stage directions to be consistent. Because we are treating stage directions as paratextual materials, they need to have a reference system that is not based on speeches (hence the Act, Scene, Stage Direction model). `<ptr>` links currently link to the correct thing, but just give a number (and we would rather have something that is similar to the ASSp links so that readers have an idea of where they will be going when they click on a link). Down the line, we would like to flesh ASSD out a bit more as Janelle described above, but for now giving an Act, Scene, Stage Direction number will help keep consistent formatting and let readers know generally what they're clicking on.

**Comment by martindholmes** [comment]
@LEMDO-PM @JanelleJenstad I guess what's being proposed, then, is a parallel system to ASSP, especially for stage directions. So we don't make any use of the existing ASSP handling (which is convoluted enough already), and we just process the xml:id attribute of any stage direction pointed at into the format needed. This only happens for `<ptr type="localCit">`. But what do we do with those pointers in the PDF? HW has one, which we can use for testing; it appears in a textual note, but that doesn't get into the PDF. None of the other print texts has an example pointer. I would need at least one of those in order to implement and test handling in the PDF, once we decide what is supposed to happen -- maybe we would just make it A1 S2 and omit the SD bit that depends on numbering that won't be there?

**Comment by martindholmes** [comment]
As of rev 21295, SDs are handled as part of the ASSP structure in both web and PDF views, and output "...SD5" style output. This means that meaningless SD numbers are being shown in the links, but there's no matching surfaced number on the web page or in the PDF itself. There are various ways we could surface the numbers in the HTML (a parenthetical number in white-on-blue, for instance, following or preceding the SD), but no realistic way to surface them in the PDF, so it might be best simply to suppress that part of the citation in the PDF view. At any rate, something has to be done about this before any release. 
Thanks @LEMDO-PM for help with this.

**Comment by martindholmes** [comment]
@LEMDO-PM @JanelleJenstad I need your guidance on how to render these things in the HTML and the PDF. This is a release-blocker -- the last one that's on my plate.

**Comment by JanelleJenstad** [comment]
@martindholmes For the PDFs, we would like a generated appendix (entitled **Stage Directions**) as we suggested in https://github.com/projectLEMDO/lemdoIssues/issues/136#issuecomment-2702323200 above. 

![Image](https://github.com/user-attachments/assets/7b7db64e-a135-4239-b59d-3d346e55d145)

For the HTML, it would be nice to have a page rendering the stage directions in tabular form (perhaps with links back to the modern text, as we do with collations and annotations?)

We think this is an in-person discussion. When you get back to this ticket, let's talk it through.

**Comment by JanelleJenstad** [comment]
Tables are not essential if they are a challenge in LaTeX. I like to use them in .docx and .odt files (and make the borders invisible) because things align and wrap so nicely in tables. There are other ways to lay out the information.

**Comment by martindholmes** [comment]
Working on this now; I have tables appearing, but they'll need a lot of work to make them attractive.

**Comment by martindholmes** [comment]
This is working for AHDM in rev 21709. H5 still needs stage-direction ids to be assigned. I'll need to do extra tweaking for HW, because there are two plays, so the table captions will have to include the play number. We may also need to tweak the column width when we have plays with acts being rendered.

**Comment by LEMDO-PM** [comment]
Fantastic! Thank you!

**Comment by martindholmes** [comment]
@JanelleJenstad @LEMDO-PM : We now have both location hints (After Sc 15 Sp 1) and page-numbers in the PDF tables; can you decide whether you want to keep both, or just go with the page-numbers?

**Comment by LEMDO-PM** [comment]
@martindholmes We would like to keep both, but please make these formatting changes:
- Add a semi-colon after page number
- Make "Before" and "After" have a lowercase at the beginning
- Align text in column three to the left
- Halve the vertical white space between the tables

So it would be "p. 3; before Sc1 Sp1"

**Comment by martindholmes** [comment]
Completed in rev 22102. Closing this.
